ssl/ssl_sess.c

```diff
@@ -44,6 +44,15 @@ static void SSL_SESSION_list_remove(SSL_CTX *ctx, SSL_SESSION *s);
 static void SSL_SESSION_list_add(SSL_CTX *ctx, SSL_SESSION *s);
 static int remove_session_lock(SSL_CTX *ctx, SSL_SESSION *c, int lck);
 
+/*
+ * TODO(TLS1.3): SSL_get_session() and SSL_get1_session() are problematic in
+ * TLS1.3 because, unlike in earlier protocol versions, the session ticket
+ * may not have been sent yet even though a handshake has finished. The session
+ * ticket data could come in sometime later...or even change if multiple session
+ * ticket messages are sent from the server. We need to work out how to deal
+ * with this.
```

Comment:
Yup, that's the behavior I'm thinking. (The internal session cache more-or-less does nothing as a client in OpenSSL anyway. Clients never look up sessions by session ID. It just wastes memory. We completely disconnected it on the client in BoringSSL.)

I don't think `SSL_get1_session` would work. Tickets are discovered post-handshake, so the consumer needs the callback to even know there's a session to stash. At which point, they may as well get the session out of the callback. Existing non-callback code likely samples the session immediately after the handshake, but we don't have a ticket at that point yet.